src/sherpa_ai/agents/base.py

```diff
@@ -105,31 +105,48 @@ def run(self):
         return result
 
     def validate_output(self):
-        last_failed_validation = None
-
-        for _ in range(self.validation_steps):
-            result = self.synthesize_output()
-            self.belief.update_internal(EventType.result, self.name, result)
-            is_valid = True
-            for validation in self.validations:
-                validation_result = validation.process_output(result, self.belief)
-
-                if not validation_result.is_valid:
-                    is_valid = False
+        """
+        Validate the synthesized output through a series of validation steps.
+
+        This method iterates through each validation in the 'validations' list, and for each validation,
+        it performs 'validation_steps' attempts to synthesize output using 'synthesize_output' method.
+        If the output doesn't pass validation, feedback is incorporated into the belief system.
+
+        If a validation fails after all attempts, the error messages from the last failed validation
+        are appended to the final result.
+
+        Returns:
+            str: The synthesized output after validation.
+        """
+        last_failed_validation = []
+        result = ""
+        for validation in self.validations:
+            for count in range(self.validation_steps):
+                result = self.synthesize_output()
+                self.belief.update_internal(EventType.result, self.name, result)
+
+                validation_result = validation.process_output(
+                    text=result, belief=self.belief, iteration_count=count
+                )
+                if validation_result.is_valid:
+                    break
+                else:
                     self.belief.update_internal(
                         EventType.feedback,
                         self.feedback_agent_name,
                         validation_result.feedback,
                     )
-                    last_failed_validation = validation
```

Comment:
I think here, until the last iteration, we should break on the first validation error and only run all validation steps for the last iteration so that we don't need to waste doing validation when we know that the results will not be used afterwards